src/lava/proc/sdn/models.py

```diff
@@ -0,0 +1,296 @@
+# Copyright (C) 2021 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause
+# See: https://spdx.org/licenses/
+
+import numpy as np
+from typing import Any, Dict
+
+from lava.magma.core.sync.protocols.loihi_protocol import LoihiProtocol
+from lava.magma.core.model.py.ports import PyInPort, PyOutPort
+from lava.magma.core.model.py.type import LavaPyType
+from lava.magma.core.resources import CPU
+from lava.magma.core.decorator import implements, requires, tag
+from lava.magma.core.model.py.model import PyLoihiProcessModel
+from lava.proc.sdn.process import Sigma, Delta, SigmaDelta, ACTIVATION_MODE
+
+
+def ReLU(x: np.ndarray) -> np.ndarray:
+    """ReLU activation implementation
+
+    Parameters
+    ----------
+    x : np.ndarray
+        input array
+
+    Returns
+    -------
+    np.ndarray
+        output array
+    """
+    return np.maximum(x, 0)
+
+
+class AbstractSigmaModel(PyLoihiProcessModel):
+    a_in = None
+    a_out = None
+
+    sigma = None
+
+    def sigma_dynamics(self, a_in_data: np.ndarray) -> np.ndarray:
+        """Sigma decoding dynamics method
+
+        Parameters
+        ----------
+        a_in_data : np.ndarray
+            Input data
+
+        Returns
+        -------
+        np.ndarray
+            decoded data
+        """
+        return a_in_data + self.sigma
+
+
+class AbstractDeltaModel(PyLoihiProcessModel):
+    a_in = None
+    s_out = None
+
+    vth = None
+    residue = None
+    error = None
+    wgt_exp = None
+    state_exp = None
+    cum_error = None
+
+    def delta_dynamics(self, act_data: np.ndarray) -> np.ndarray:
+        """Delta encodind dynamics method
+
+        Parameters
+        ----------
+        act_data : np.ndarray
+            data to be encoded
+
+        Returns
+        -------
+        np.ndarray
+            delta encoded data
+        """
+        delta = act_data - self.act + self.residue
+
+        if self.cum_error:
+            self.error += delta
+            s_out = np.where(
+                np.abs(self.error) >= self.vth,
+                delta, 0
+            )
+            self.error *= 1 - (np.abs(s_out) > 0)
+        else:
+            s_out = np.where(
+                np.abs(delta) >= self.vth,
+                delta, 0
+            )
+        self.residue = delta - s_out
+        return s_out
+
+
+class AbstractSigmaDeltaModel(AbstractSigmaModel, AbstractDeltaModel):
+    a_in = None
+    s_out = None
+
+    vth = None
+    sigma = None
+    act = None
+    residue = None
+    error = None
+    bias = None
+    wgt_exp = None
+    state_exp = None
+    cum_error = None
```

Comment:
Perhaps this is for clarity but we are redefining these here, as they are also defined in the base classes.

Reply:
Yes this is for clarity to keep track of vars that needs to be overwritten in the eventual leaf implementation.